# Exploring Ebay Car Sales Data

## Introduction
The aim of this project is to clean the data and analyze the included used car listings from eBay Kleinanzeigen, a classifieds section of the German eBay website.The original dataset was scraped and uploaded to Kaggle by user orgesleka is no longer available, but you can now find it [here](https://data.world/data-society/used-cars-data)

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

## Opening and Exploring data
Lets start by opening data set and continue with exploring the data.

In [81]:
# Import the pandas and NumPy libraries
import pandas as pd
import numpy as np

#Read the autos.csv CSV file into pandas
autos = pd.read_csv("/Users/sahithi/Documents/Python Projects/Exploring Ebay Car Sales Data/autos.csv", encoding='Latin-1')

In [82]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [83]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

### Observations:
Right away, we can see that the following columns have missing (null) values:

- vehicleType
- gearbox
- model
- fuelType
- notRepairedDamage

Also, it is worth noting that the columns are using camelcase instead of snakecase. Let us modify the columns to Python's preferred snakecase.


## Column Renaming

In [84]:
# to print an array of the existing column names.
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [85]:
#modified new columns
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [86]:
# assigned new column names
autos.columns = new_columns
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

After modifying the column headers, let us begin exploring the data set.

Some things to look out for:

Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
Examples of numeric data stored as text which can be cleaned and converted.
Commonly used methods include:

- describe() method
- value_counts() method
- isnull() method

In [87]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

From the table above, we can conclude:

. The seller and offer_type columns have only 2 unique values. Therefore, those columns contain virtually no useful information for analysis.

. The nr_of_pictures column is irrelevant to us, as all rows present only the 0.0 value.

## Deleting irrelevent columns

### seller, offer_type and nr_of_pictures

In [88]:
# drop the columns
autos.drop(columns=['seller', 'offer_type','nr_of_pictures'], inplace= True)

In [89]:
autos.shape

(371528, 17)

## Describing numeric columns
Now we can find out how many unique values numeric columns have. It is necessary to make sure that in many rows they have more than one value - otherwise these columns will be viewed as statistically insignificant and, consequently, will be removed from the dataframe

In [90]:
# Calculating the number of unique values for each numeric column
print('Unique values for numeric columns:')
print('price:', *autos['price'].unique().shape)
print('registration_year:', *autos['registration_year'].unique().shape)
print('kilometer:', *autos['kilometer'].unique().shape)
print('power_ps:', *autos['power_ps'].unique().shape)
print('registration_month:', *autos['registration_month'].unique().shape)
print('postal_code:', *autos['postal_code'].unique().shape)

Unique values for numeric columns:
price: 5597
registration_year: 155
kilometer: 13
power_ps: 794
registration_month: 13
postal_code: 8150


As we can see, all the four numeric columns have a plenty of unique values - that proves that they statistically significant and that they cannot be dropped

## Data Exploration

Before removing the outliers from 'kilometer' and 'price' columns, we need to describe these columns (with outliers included) in order to be able to compare them with themselves, but without outliers

In [96]:
autos['price'].value_counts()

0         10778
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price, Length: 5597, dtype: int64

In [97]:
autos['price'].value_counts().sort_index(ascending = True)

0             10778
1              1189
2                12
3                 8
4                 1
              ...  
32545461          1
74185296          1
99000000          1
99999999         15
2147483647        1
Name: price, Length: 5597, dtype: int64

In [98]:
autos['price'].value_counts().sort_index().tail(53)

350000         4
368000         1
370000         2
390000         1
395000         1
399997         1
420000         1
440000         1
445000         1
466000         1
485000         1
487000         1
488997         1
500000         2
517895         1
579000         1
585000         1
599000         1
600000         2
619000         1
650000         1
700000         1
725000         1
745000         2
780000         1
820000         1
849000         1
911911         1
999990         1
999999        13
1000000        5
1010010        1
1111111        2
1234566        1
1250000        2
1300000        1
1600000        2
2795000        1
2995000        1
3890000        1
3895000        1
9999999        3
10000000       8
10010011       1
11111111      10
12345678       9
14000500       1
27322222       1
32545461       1
74185296       1
99000000       1
99999999      15
2147483647     1
Name: price, dtype: int64

In [99]:
autos[autos['price'] > 350000].sort_values('price')

date_crawled  \
336623  2016-03-17 09:49:55   
168786  2016-04-02 23:53:46   
195299  2016-03-31 00:55:12   
328614  2016-03-10 15:47:17   
82700   2016-04-05 11:55:33   
303287  2016-03-23 16:54:56   
10649   2016-04-05 09:36:18   
360768  2016-03-21 17:39:22   
201679  2016-04-01 17:58:14   
330117  2016-04-01 15:56:24   
155596  2016-03-23 11:55:37   
337851  2016-03-24 18:59:49   
139405  2016-03-22 15:37:43   
77185   2016-03-21 19:50:05   
119382  2016-03-21 19:57:59   
118058  2016-04-07 04:36:19   
1846    2016-03-10 22:47:05   
136495  2016-03-27 17:39:13   
365461  2016-03-31 18:53:48   
161773  2016-03-12 16:39:54   
26327   2016-03-21 19:43:54   
137856  2016-03-19 01:36:24   
169098  2016-03-08 10:36:17   
100851  2016-03-12 18:46:35   
111647  2016-03-17 00:57:28   
118143  2016-03-20 12:30:26   
344196  2016-03-05 17:25:25   
163190  2016-03-27 18:53:03   
78265   2016-03-31 19:55:47   
194217  2016-04-03 15:52:51   
85047   2016-03-28 12:45:00   
359083  2016-03-29 11:38:54   
131785  2016-04-03 18:58:58   
332531  2016-03-20 02:36:23   
38445   2016-03-21 19:53:52   
118233  2016-03-09 21:56:45   
69726   2016-03-17 15:58:57   
312909  2016-03-17 09:53:08   
159974  2016-04-01 16:36:57   
301198  2016-04-04 09:36:42   
155418  2016-03-27 20:52:14   
169436  2016-04-01 12:56:01   
141004  2016-04-02 13:47:00   
174617  2016-03-14 21:40:52   
169656  2016-03-21 20:45:24   
45833   2016-03-15 10:53:11   
312109  2016-04-05 11:50:18   
16889   2016-03-15 16:45:45   
306469  2016-03-17 12:06:35   
53507   2016-03-11 18:57:57   
173279  2016-03-07 21:36:25   
57594   2016-03-25 09:56:28   
106651  2016-03-11 22:54:08   
167750  2016-03-22 12:54:19   
20143   2016-03-18 00:55:53   
78480   2016-03-10 20:43:12   
316200  2016-04-04 11:53:31   
367041  2016-03-24 19:00:03   
333699  2016-03-12 03:36:18   
45055   2016-04-02 00:06:21   
41980   2016-03-11 09:37:17   
364171  2016-04-04 21:25:21   
366861  2016-03-21 10:53:06   
347435  2016-03-30 15:50:46   
91106   2016-03-30 19:46:05   
21467   2016-03-19 19:45:02   
363434  2016-03-23 06:36:20   
358868  2016-03-16 16:49:53   
121793  2016-03-05 20:59:54   
281833  2016-03-19 08:50:50   
54295   2016-03-08 11:51:40   
171824  2016-03-29 23:52:57   
176070  2016-04-02 14:43:10   
196306  2016-04-01 01:59:20   
361451  2016-04-03 14:36:47   
318143  2016-03-21 13:57:51   
332028  2016-03-31 18:39:53   
177744  2016-03-31 16:37:00   
185227  2016-03-14 16:52:51   
127769  2016-03-28 12:51:51   
338415  2016-03-26 22:49:39   
54507   2016-03-09 11:40:15   
90853   2016-04-01 05:36:16   
210657  2016-03-12 21:50:57   
14663   2016-04-03 06:03:56   
287835  2016-03-26 00:56:36   
20746   2016-03-08 15:54:52   
217547  2016-03-11 10:40:52   
241847  2016-03-31 18:56:54   
266124  2016-04-03 23:36:16   
273145  2016-03-09 15:45:47   
299872  2016-03-14 10:06:23   
310924  2016-03-14 05:36:22   
39625   2016-03-08 23:53:51   
279937  2016-04-03 11:56:00   
87799   2016-03-08 20:39:05   
183361  2016-03-26 11:51:20   
120919  2016-03-25 12:43:02   
251980  2016-03-21 22:25:19   
150909  2016-04-02 22:41:54   
77520   2016-03-22 14:58:27   
366653  2016-03-20 10:59:52   
362782  2016-03-06 10:38:42   
127820  2016-03-14 17:49:54   
56973   2016-03-18 18:52:27   
108987  2016-03-29 00:38:37   
69747   2016-03-20 10:59:32   
325589  2016-03-27 21:55:14   
160530  2016-03-29 12:58:27   
189737  2016-04-03 12:41:33   
280059  2016-03-05 20:49:28   
268173  2016-04-01 09:37:19   
246635  2016-03-25 11:55:36   
139052  2016-03-14 12:49:48   
129969  2016-03-11 08:37:57   

                                                     name       price  \
336623                 Mercedes_Benz_SLR_McLaren_Roadster      368000   
168786                                    Ferrari_512_BBi      370000   
195299                  Porsche_911_GT2_RS_NUR_500_Stueck      370000   
328614                                    BMW_Z8_roadster      390000   
82700   Ferrari_246_GTS__Certificate_Classiche_von_Fe

### Removing outliers from 'price'

In [136]:
# Removing the outliers from the 'price' column
autos = autos[(((autos['price'] >= 1) & (autos['price'] <= 350000 )) | ((autos["name"] == 'Ferrari' ) & (autos["name"] == 'Maserati')))]

autos['price'].value_counts().sort_index()

1        922
2          6
3          6
5         19
7          1
        ... 
16199      6
16200     65
16215      1
16250     28
16270      1
Name: price, Length: 2989, dtype: int64

### kilometer

In [144]:
# See how many unique values are in the column

autos['kilometer'].unique().shape

(4,)

In [148]:
autos['kilometer'].describe()

count    286642.000000
mean     142345.504148
std       16467.169306
min       90000.000000
25%      150000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64

In [149]:
autos['kilometer'].value_counts()

150000    227618
125000     34564
100000     13760
90000      10700
Name: kilometer, dtype: int64

The odometer_km column seems to have no major issues:

The min value is 90000 km, present in 10700 entries.
The max value is 150000 km, which is also the most frequent value, occuring in 227618 entries.
The mean is 142345.5 km, with a standard deviation of 16467.1 km.
Most entries have a value superior to 900000 km. Since we are working with used cars, this seems reasonable. Therefore, we will not remove any entries based on unrealistically high or low kilometer values.

### Comparing the columns with and without outliers
Now we can compare the values of the columns with outliers included with the values of the same columns without outliers:

In [108]:
# Printing the table with outliers
outliers_table

kilometer         price
count  371528.000000  3.715280e+05
mean   125618.688228  1.729514e+04
std     40112.337051  3.587954e+06
min      5000.000000  0.000000e+00
25%    125000.000000  1.150000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  2.147484e+09

In [109]:
# Observing the values left in 'kilometer' and 'price' after removing outliers
autos[['kilometer', 'price']].describe()

kilometer          price
count  295620.000000  295620.000000
mean   142410.594682    3563.428415
std     16413.471333    3532.760660
min     90000.000000       0.000000
25%    150000.000000     999.000000
50%    150000.000000    2300.000000
75%    150000.000000    4999.000000
max    150000.000000   16270.000000

After the outliers were removed, 'kilometer' has experienced some changes. Its smallest values were outliers and their removal, consequently lead to smaller standard deviation (16413 instead of 40112) and bigger minimum value (90000 instead of 5000).

The new 'price' column has changed significantly as well: its standard deviation is now equal to 3532 instead of 3587; its maximum value is now 16270 instead of 2.147484e+09

### Registaration year

In [153]:
# see how many unique values
autos['registration_year'].unique().shape

(89,)

In [155]:
autos['registration_year'].describe()

count    286642.000000
mean       2002.539391
std           7.311772
min        1000.000000
25%        1999.000000
50%        2002.000000
75%        2006.000000
max        3000.000000
Name: registration_year, dtype: float64

In [157]:
autos['registration_year'].value_counts().sort_index()

1000        1
1400        1
1600        1
1910       24
1928        1
1930        1
1931        1
1932        1
1934        1
1936        1
1937        1
1938        1
1940        1
1941        1
1942        2
1943        1
1945        2
1946        1
1950        8
1951        3
1952        3
1953        9
1954        4
1955        4
1956        6
1957        8
1958        7
1959       10
1960       50
1961        8
1962       18
1963       19
1964       20
1965       40
1966       50
1967       41
1968       46
1969       58
1970      133
1971       77
1972       95
1973       86
1974       85
1975       60
1976       75
1977      102
1978      158
1979      181
1980      294
1981      170
1982      211
1983      293
1984      272
1985      478
1986      369
1987      507
1988      757
1989     1025
1990     2041
1991     2370
1992     2610
1993     3023
1994     4355
1995     8258
1996     9770
1997    13314
1998    16293
1999    21012
2000    21555
2001    18589
2002    17669
2003  

Since the car couldn't be registered before 1900 and after 2016 which is year where this listings were added we will remove this outliers.

In [213]:
autos = autos[autos['registration_year'].between(1900,2019)]

In [214]:
autos['registration_year'].value_counts().sort_index()

1910       24
1928        1
1930        1
1931        1
1932        1
1934        1
1936        1
1937        1
1938        1
1940        1
1941        1
1942        2
1943        1
1945        2
1946        1
1950        8
1951        3
1952        3
1953        9
1954        4
1955        4
1956        6
1957        8
1958        7
1959       10
1960       50
1961        8
1962       18
1963       19
1964       20
1965       40
1966       50
1967       41
1968       46
1969       58
1970      133
1971       77
1972       95
1973       86
1974       85
1975       60
1976       75
1977      102
1978      158
1979      181
1980      294
1981      170
1982      211
1983      293
1984      272
1985      478
1986      369
1987      507
1988      757
1989     1025
1990     2041
1991     2370
1992     2610
1993     3023
1994     4355
1995     8258
1996     9770
1997    13314
1998    16293
1999    21012
2000    21555
2001    18589
2002    17669
2003    18346
2004    18031
2005    19525
2006  

## Date values

There are 5 columns that should represent date values, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a astype datetime representation so we can understand it quantitatively. The other two registration_month and registration_year columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [161]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21
5  2016-04-04 17:36:23  2016-04-04 00:00:00  2016-04-06 19:17:07

We can see that the date is in a format YYYY-MM-DD HH-MM-SS. We are going to use pd.to_datetime to change astype to datetime and we will change format to YYYY-MM-DD.

### date_crawled

In [162]:
# change dtype to datetime
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'], format = "%Y-%m-%d %H:%M:%S")
# change format to YYYY-MM-DD
date_crawled = autos['date_crawled'].dt.strftime('%Y-%m-%d')

In [163]:
date_crawled.value_counts(normalize = True, dropna = False)

2016-04-03    0.038135
2016-04-04    0.037437
2016-03-14    0.036492
2016-03-12    0.036248
2016-03-21    0.036066
2016-03-20    0.036052
2016-03-07    0.035690
2016-03-19    0.035257
2016-03-28    0.034967
2016-04-02    0.034538
2016-03-09    0.034252
2016-03-29    0.034012
2016-04-01    0.033959
2016-03-08    0.033823
2016-03-30    0.033621
2016-03-15    0.033617
2016-03-10    0.033122
2016-03-22    0.033059
2016-03-11    0.032951
2016-03-25    0.032616
2016-03-17    0.032142
2016-03-23    0.032128
2016-03-26    0.031744
2016-03-31    0.031695
2016-03-16    0.030373
2016-03-27    0.030080
2016-03-24    0.029884
2016-03-05    0.025468
2016-03-13    0.015420
2016-03-06    0.014367
2016-03-18    0.013306
2016-04-05    0.012744
2016-04-06    0.003157
2016-04-07    0.001577
Name: date_crawled, dtype: float64

### 7.2 ad_created

In [170]:
# change dtype to datetime
autos['ad_created'] = pd.to_datetime(autos['ad_created'], format = "%Y-%m-%d %H:%M:%S")
# change format to YYYY-MM-DD
ad_created = autos['ad_created'].dt.strftime('%Y-%m')

In [171]:
ad_created.value_counts(normalize = True, dropna = False)

2016-03    0.838399
2016-04    0.160370
2016-02    0.001022
2016-01    0.000150
2015-12    0.000024
2015-11    0.000017
2015-09    0.000007
2015-08    0.000003
2014-03    0.000003
2015-03    0.000003
Name: ad_created, dtype: float64

The eBay listings were created between March 2014 and April 2016. But 99% of the ads were created in March and April 2016. Most adds were created in March 2016 - 83.8%

### last_seen

In [172]:
# change dtype to datetime
autos['last_seen'] = pd.to_datetime(autos['last_seen'], format = "%Y-%m-%d %H:%M:%S")
# change format to YYYY-MM-DD
last_seen= autos['last_seen'].dt.strftime('%Y-%m-%d')

In [174]:
last_seen.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001399
2016-03-06    0.004410
2016-03-07    0.005701
2016-03-08    0.008676
2016-03-09    0.010512
2016-03-10    0.012211
2016-03-11    0.013766
2016-03-12    0.024990
2016-03-13    0.008928
2016-03-14    0.012891
2016-03-15    0.017294
2016-03-16    0.017398
2016-03-17    0.030551
2016-03-18    0.007333
2016-03-19    0.017468
2016-03-20    0.021194
2016-03-21    0.021288
2016-03-22    0.021571
2016-03-23    0.018703
2016-03-24    0.020259
2016-03-25    0.020064
2016-03-26    0.016823
2016-03-27    0.017656
2016-03-28    0.023434
2016-03-29    0.024302
2016-03-30    0.024442
2016-03-31    0.024896
2016-04-01    0.024784
2016-04-02    0.026008
2016-04-03    0.026301
2016-04-04    0.026033
2016-04-05    0.120978
2016-04-06    0.205475
2016-04-07    0.122262
Name: last_seen, dtype: float64

The crawler saw this ads between 5th of March and 7th of April 2016

## Translate to English

We will translate data from this columns: vehicle_type, gearbox, model, fuel_type, brand and unrepaired_damage from german to english. First we will check unique values for every column, translate them and than map the values to their english counterparts.



### vehicle_type

In [185]:
vehicle_type = pd.unique(autos['vehicle_type'])

In [186]:
vehicle_type

array([nan, 'suv', 'kleinwagen', 'limousine', 'cabrio', 'kombi', 'bus',
       'coupe', 'andere'], dtype=object)

In [180]:
new_vt = ['limousine', 'small_car', 'combi', 'bus', 'convertible', 'coupe', 'suv', 'other']

In [187]:
vehicle_type = new_vt

In [188]:
vehicle_type

['limousine',
 'small_car',
 'combi',
 'bus',
 'convertible',
 'coupe',
 'suv',
 'other']

### gearbox

In [189]:
gearbox = pd.unique(autos['gearbox'])

In [190]:
gearbox

array(['manuell', 'automatik', nan], dtype=object)

In [191]:
new_gearbox = ['manual', 'automatic', 'nan']

In [192]:
gearbox=new_gearbox

In [193]:
gearbox

['manual', 'automatic', 'nan']

### fuel_type

In [194]:
fuel_type = pd.unique(autos['fuel_type'])

In [195]:
fuel_type

array(['benzin', 'diesel', nan, 'lpg', 'andere', 'hybrid', 'cng',
       'elektro'], dtype=object)

In [196]:
new_ft = ['petrol', 'diesel', 'lpg', 'cng', 'hybrid', 'electric', 'other']

In [197]:
fuel_type = new_ft

In [198]:
fuel_type

['petrol', 'diesel', 'lpg', 'cng', 'hybrid', 'electric', 'other']

### unrepaired_damage

In [199]:
unrepaired_damage = pd.unique(autos['unrepaired_damage'])

In [200]:
unrepaired_damage

array([nan, 'nein', 'ja'], dtype=object)

In [201]:
new_urd = ['nan', 'no', 'yes']

In [202]:
unrepaired_damage = new_urd

In [203]:
unrepaired_damage

['nan', 'no', 'yes']

### model

In [204]:
print(pd.unique(autos['model']))

['golf' 'grand' 'fabia' '3er' '2_reihe' '3_reihe' 'passat' 'polo' 'twingo'
 'c_max' 'a_klasse' 'scirocco' '5er' 'meriva' 'arosa' 'andere' 'c4'
 'punto' 'e_klasse' 'clio' 'kangoo' 'one' '1er' 'b_klasse' 'signum'
 'astra' nan 'a8' 'jetta' 'fiesta' 'c_klasse' 'vito' 'sprinter' '156'
 'escort' 'forester' 'xc_reihe' 'scenic' 'a4' 'ka' 'insignia'
 'transporter' 'combo' 'focus' 'tt' 'civic' 'corsa' 'a6' 'jazz' 'omega'
 'slk' '7er' '80' '147' '100' 'z_reihe' 'sportage' 'v40' 'touran' 'getz'
 'a3' 'almera' 'megane' 'fortwo' 'lupo' 'r19' 'zafira' 'caddy' 'mondeo'
 'cordoba' 'impreza' 'a1' 'vectra' 'tiguan' 'i_reihe' 'espace' 'sharan'
 'panda' 'ceed' 'ibiza' 'octavia' 'rx_reihe' 'modus' 'matiz' 'beetle'
 'touareg' 'logan' '6_reihe' 'spider' 'cuore' 's_max' 'a2' 'c3' 's_klasse'
 '1_reihe' 'eos' 'avensis' 'viano' 'sl' 'kaefer' 'santa' 'fox' 'cooper'
 'leon' 'laguna' 'ptcruiser' 'clk' 'primera' '4_reihe' 'galaxy' 'x_reihe'
 'micra' '159' 'qashqai' 'carisma' 'accord' 'corolla' 'lanos' 'phaeton'
 'ver

Since we have alot unique values, we will translate only andere to other.

In [205]:
autos['model'] = autos['model'].str.replace('andere', 'other')
print(pd.unique(autos['model']))

['golf' 'grand' 'fabia' '3er' '2_reihe' '3_reihe' 'passat' 'polo' 'twingo'
 'c_max' 'a_klasse' 'scirocco' '5er' 'meriva' 'arosa' 'other' 'c4' 'punto'
 'e_klasse' 'clio' 'kangoo' 'one' '1er' 'b_klasse' 'signum' 'astra' nan
 'a8' 'jetta' 'fiesta' 'c_klasse' 'vito' 'sprinter' '156' 'escort'
 'forester' 'xc_reihe' 'scenic' 'a4' 'ka' 'insignia' 'transporter' 'combo'
 'focus' 'tt' 'civic' 'corsa' 'a6' 'jazz' 'omega' 'slk' '7er' '80' '147'
 '100' 'z_reihe' 'sportage' 'v40' 'touran' 'getz' 'a3' 'almera' 'megane'
 'fortwo' 'lupo' 'r19' 'zafira' 'caddy' 'mondeo' 'cordoba' 'impreza' 'a1'
 'vectra' 'tiguan' 'i_reihe' 'espace' 'sharan' 'panda' 'ceed' 'ibiza'
 'octavia' 'rx_reihe' 'modus' 'matiz' 'beetle' 'touareg' 'logan' '6_reihe'
 'spider' 'cuore' 's_max' 'a2' 'c3' 's_klasse' '1_reihe' 'eos' 'avensis'
 'viano' 'sl' 'kaefer' 'santa' 'fox' 'cooper' 'leon' 'laguna' 'ptcruiser'
 'clk' 'primera' '4_reihe' 'galaxy' 'x_reihe' 'micra' '159' 'qashqai'
 'carisma' 'accord' 'corolla' 'lanos' 'phaeton' 'verso

### brand

In [206]:
print(pd.unique(autos['brand']))

['volkswagen' 'jeep' 'skoda' 'bmw' 'peugeot' 'mazda' 'renault' 'ford'
 'mercedes_benz' 'opel' 'seat' 'citroen' 'fiat' 'mini' 'sonstige_autos'
 'audi' 'honda' 'alfa_romeo' 'subaru' 'volvo' 'mitsubishi' 'kia' 'hyundai'
 'nissan' 'smart' 'suzuki' 'lancia' 'toyota' 'chevrolet' 'dacia'
 'daihatsu' 'trabant' 'chrysler' 'jaguar' 'daewoo' 'porsche' 'rover'
 'saab' 'land_rover' 'lada']


In [207]:
autos['brand'] = autos['brand'].str.replace('sonstige_autos', 'other')
print(pd.unique(autos['brand']))

['volkswagen' 'jeep' 'skoda' 'bmw' 'peugeot' 'mazda' 'renault' 'ford'
 'mercedes_benz' 'opel' 'seat' 'citroen' 'fiat' 'mini' 'other' 'audi'
 'honda' 'alfa_romeo' 'subaru' 'volvo' 'mitsubishi' 'kia' 'hyundai'
 'nissan' 'smart' 'suzuki' 'lancia' 'toyota' 'chevrolet' 'dacia'
 'daihatsu' 'trabant' 'chrysler' 'jaguar' 'daewoo' 'porsche' 'rover'
 'saab' 'land_rover' 'lada']


## Analyze data

### Mean price by brand.

First we will calculate the mean price for every brand. We can use aggregation to understand the brand column.

In [216]:
autos['brand'].value_counts().sort_index

<bound method Series.sort_index of volkswagen       60625
opel             31649
bmw              30816
mercedes_benz    26319
audi             23958
ford             19291
renault          14085
peugeot           8598
fiat              6750
seat              5029
mazda             4443
skoda             3913
citroen           3781
nissan            3645
toyota            3353
smart             3055
volvo             2806
mitsubishi        2386
hyundai           2206
honda             2147
alfa_romeo        1872
mini              1789
kia               1609
suzuki            1489
other             1459
chrysler          1189
chevrolet         1001
subaru             605
daihatsu           596
jeep               533
porsche            476
saab               475
daewoo             430
land_rover         423
jaguar             411
dacia              404
rover              397
lancia             372
trabant             95
lada                70
Name: brand, dtype: int64>

We have 40 diffrent car brands in our data. The most popular brand is Volkswagen. We will continue to analyze only for the 10 most popular brands.

In [217]:
#names of Top10 brands
top10 = autos['brand'].value_counts().index[:10]

In [218]:
top10

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')

In [221]:
# create a dictionary
mean_price = {}
for brand in top10:
    selected_rows = autos[autos['brand']== brand]
    price = selected_rows['price'].mean()
    mean_price[brand] = int(price)
    
mean_price = pd.Series(mean_price).sort_values(ascending=False)

In [224]:
brand_price = pd.DataFrame(mean_price, columns =['mean_price'])
brand_price

mean_price
bmw                  5526
audi                 5365
mercedes_benz        5062
volkswagen           3670
seat                 2929
ford                 2500
peugeot              2479
opel                 2237
fiat                 1939
renault              1820

We can see that in the top 10 brands, there's a distinct price gap.

- Audi, Mercedes Benz and BMW are more expensive
- Fiat, Peugeot, Reanult and Opel are less expensive
- Volkswagen, Ford and Seat are in between


### Mean car mileage by brand.

In [229]:
# create a dictionary
mean_mileage = {}
for brand in top10:
    selected_rows = autos[autos['brand']==brand] 
    mileage = selected_rows['kilometer'].mean()
    mean_mileage[brand] = int(mileage)
    
mean_mileage = pd.Series(mean_mileage).sort_values(ascending = False)

In [230]:
brand_mileage = pd.DataFrame(mean_mileage, columns =[' mean_mileage'])
brand_mileage

mean_mileage
audi                  145784
bmw                   145227
mercedes_benz         144782
volkswagen            143615
opel                  141920
renault               141157
seat                  140580
ford                  140357
peugeot               139641
fiat                  137940

Audi cars has the highest mileage. Now we will check if the price of each brand depends on car mileage.

In [231]:
brand_price['mean_mileage']= brand_mileage
brand_price

mean_price  mean_mileage
bmw                  5526        145227
audi                 5365        145784
mercedes_benz        5062        144782
volkswagen           3670        143615
seat                 2929        140580
ford                 2500        140357
peugeot              2479        139641
opel                 2237        141920
fiat                 1939        137940
renault              1820        141157

In [232]:
brand_price["mean_mileage"].describe()

count        10.000000
mean     142100.300000
std        2634.402992
min      137940.000000
25%      140412.750000
50%      141538.500000
75%      144490.250000
max      145784.000000
Name: mean_mileage, dtype: float64

There are no drastic differences in the average mileage between the top 10 brands. All brands display a mean mileage over 110000 km, and the standard deviation of the series is below 5000 km.
The brand with the highest average mileage is Audi (145784 km) and the one with the lowest is fiat (137940 km). The correlation between these two columns ('mean_price' and 'mean_mileage') is equal to 43% - that is considered to be a weak correlation. Then it can be concluded that these two characteristics ('mean_price' and 'mean_mileage') are weakly correlated; consequently, the changes in one column are not likely to influence the other column (and the other way around).

## Conclusion
We performed several data cleaning tasks before analyzing the dataset, such as:

- removing 3 irrelevant columns (seller, offer_type, and nr_of_pictures) that added no valuable information;
- excluding unreasonable entries in the kilometer, price and registration_year columns, which we categorized as outliers.

Regarding the analysis of the data, we have determined that:

- at the time of crawling, the majority of cars on sale in the eBay Kleinanzeigen platform were at least more than 5 years old;
- most of the used cars announced in the crawled ads were sold after less than 2 weeks after publishing;
- Among the top 10 most common brands, there were 3 luxury brands (audi, mercedes_benz and bmw, average price ~9000 dollars), middle-range brands (volkswagen and seat, average price ~5000 dollars) and cheaper brands (fiat and renault, average price ~3000 dollars);
- Mean price of a car brand and its mean mileage are weakly correlated.